File: packages/pyrx_synapse_ios/ios/Classes/PyrxSynapseHostApiImpl.swift
// PyrxSynapseHostApiImpl.swift
// pyrx_synapse_ios — iOS Pigeon HostApi implementation.
//
// Implements the Pigeon-generated `PyrxSynapseHostApi` protocol (from
// `Classes/PyrxSynapseMessages.g.swift`) by forwarding every method to
// the `Pyrx.shared` actor from PYRXSynapse 0.1.2+. Every call:
//
//   1. Decodes Pigeon-typed arguments into the native SDK's own types.
//   2. Hops into `Pyrx.shared`'s actor isolation with `Task { await … }`.
//   3. Resolves or rejects the Pigeon completion handler.
//
// Error mapping mirrors `pyrx-synapse-react-native/ios/.../PyrxSynapseImpl.swift`:
//
//   .notInitialized       → "not_initialized"
//   .invalidConfig        → "invalid_argument"
//   .network              → "network_error"
//   .keychainFailure      → "internal_error"
//   .alreadyInitialized   → "invalid_argument"
//   <anything else>       → "internal_error"
//
// The error code reaches the Dart-side caller as the `code` field of
// `FlutterError`; PR-2's `SynapseError` Dart class wraps these so app
// code can pattern-match without depending on Pigeon types directly.

import Flutter
import Foundation
import PYRXSynapse

final class PyrxSynapseHostApiImpl: PyrxSynapseHostApi {

  // MARK: Lifecycle

  func initialize(
    args: PyrxInitArgs,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let workspaceUuid = UUID(uuidString: args.workspaceId) else {
      completion(.failure(Self.flutterError(
        code: "invalid_argument",
        message: "workspaceId must be a UUID v4 string"
      )))
      return
    }

    let environment: PyrxEnvironment
    switch args.environment.lowercased() {
    case "production", "live":
      environment = .production
    case "sandbox", "staging", "test":
      environment = .sandbox
    default:
      completion(.failure(Self.flutterError(
        code: "invalid_argument",
        message: "environment must be one of: production, sandbox"
      )))
      return
    }

    let baseUrl: URL = {
      if let str = args.baseUrl, let url = URL(string: str) {
        return url
      }
      return PyrxConfig.defaultBaseUrl
    }()

    let logLevel = Self.parseLogLevel(args.logLevel) ?? .info

    let config = PyrxConfig(
      workspaceId: workspaceUuid,
      apiKey: args.apiKey,
      environment: environment,
      baseUrl: baseUrl,
      logLevel: logLevel,
      sdkVariant: "flutter"
    )

    Task {
      do {
        try await Pyrx.shared.initialize(config: config)
        completion(.success(()))
      } catch {
        completion(.failure(Self.translate(error)))
      }
    }
  }

  func setLogLevel(
    level: String,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let parsed = Self.parseLogLevel(level) else {
      completion(.failure(Self.flutterError(
        code: "invalid_argument",
        message: "level must be one of: debug, info, warning, error, none"
      )))
      return
    }
    Pyrx.shared.setLogLevel(parsed)
    completion(.success(()))
  }

  func debugInfo(
    completion: @escaping (Result<PyrxDebugInfo, Error>) -> Void
  ) {
    Task {
      let info = await Pyrx.shared.debugInfo()
      completion(.success(PyrxDebugInfo(
        sdkVersion: info.sdkVersion,
        platform: info.platform,
        initialized: info.initialized,
        workspaceId: info.workspaceId?.uuidString,
        environment: info.environment,
        baseUrl: info.baseUrl,
        logLevel: Self.logLevelString(info.logLevel),
        anonymousId: info.anonymousId,
        externalId: info.externalId,
        trackingEnabled: info.trackingEnabled,
        queueDepth: Int64(info.queueDepth),
        deviceTokenFingerprint: info.deviceTokenFingerprint
      )))
    }
  }

  // MARK: Identity

  func identify(
    externalId: String,
    traitsJson: String?,
    completion: @escaping (Result<PyrxIdentityResult, Error>) -> Void
  ) {
    let traits: [String: JSONValue]?
    do {
      traits = try Self.decodeTraits(traitsJson)
    } catch {
      completion(.failure(Self.flutterError(
        code: "invalid_argument",
        message: "traits is not valid JSON: \(error.localizedDescription)"
      )))
      return
    }

    Task {
      do {
        let result = try await Pyrx.shared.identify(externalId: externalId, traits: traits)
        completion(.success(Self.encodeIdentity(result)))
      } catch {
        completion(.failure(Self.translate(error)))
      }
    }
  }

  func alias(
    newExternalId: String,
    completion: @escaping (Result<PyrxIdentityResult, Error>) -> Void
  ) {
    Task {
      do {
        let result = try await Pyrx.shared.alias(newExternalId: newExternalId)
        completion(.success(Self.encodeIdentity(result)))
      } catch {
        completion(.failure(Self.translate(error)))
      }
    }
  }

  func logout(completion: @escaping (Result<Void, Error>) -> Void) {
    Task {
      do {
        try await Pyrx.shared.logout()
        completion(.success(()))
      } catch {
        completion(.failure(Self.translate(error)))
      }
    }
  }

  // MARK: Events

  func track(
    eventName: String,
    propertiesJson: String?,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    let properties: [String: JSONValue]?
    do {
      properties = try Self.decodeTraits(propertiesJson)
    } catch {
      completion(.failure(Self.flutterError(
        code: "invalid_argument",
        message: "properties is not valid JSON: \(error.localizedDescription)"
      )))
      return
    }

    Task {
      do {
        try await Pyrx.shared.track(eventName: eventName, properties: properties)
        completion(.success(()))
      } catch {
        completion(.failure(Self.translate(error)))
      }
    }
  }

  func screen(
    screenName: String,
    propertiesJson: String?,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    let properties: [String: JSONValue]?
    do {
      properties = try Self.decodeTraits(propertiesJson)
    } catch {
      completion(.failure(Self.flutterError(
        code: "invalid_argument",
        message: "properties is not valid JSON: \(error.localizedDescription)"
      )))
      return
    }

    Task {
      do {
        try await Pyrx.shared.screen(screenName: screenName, properties: properties)
        completion(.success(()))
      } catch {
        completion(.failure(Self.translate(error)))
      }
    }
  }

  // MARK: Push

  func requestPushPermission(
    alert: Bool,
    sound: Bool,
    badge: Bool,
    completion: @escaping (Result<PyrxPushPermissionResult, Error>) -> Void
  ) {
    var options: UNAuthorizationOptions = []
    if alert { options.insert(.alert) }
    if sound { options.insert(.sound) }
    if badge { options.insert(.badge) }

    Task {
      let status = await Pyrx.shared.requestPushPermission(options: options)
      completion(.success(PyrxPushPermissionResult(status: Self.permissionString(status))))
    }
  }

  func registerForPushNotifications(
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    DispatchQueue.main.async {
      UIApplication.shared.registerForRemoteNotifications()
      completion(.success(()))
    }
  }

  // MARK: Privacy

  func setTrackingEnabled(
    enabled: Bool,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    Task {
      await Pyrx.shared.setTrackingEnabled(enabled)
      completion(.success(()))
    }
  }

  func deleteUser(completion: @escaping (Result<Void, Error>) -> Void) {
    Task {
      do {
        try await Pyrx.shared.deleteUser()
        completion(.success(()))
      } catch {
        completion(.failure(Self.translate(error)))
      }
    }
  }

  // MARK: - In-app messaging (Phase 10 PR-2b)
  //
  // Five methods that delegate to `Synapse.InApp.*` — the public iOS
  // facade in PYRXSynapse 0.2.0. The per-placement render callback
  // we register with `Synapse.InApp.show` is a no-op: the per-message
  // dispatch into Dart happens via the `inAppMessageReceived` envelope
  // on the shared event stream (PyrxEventStreamHandler), so Flutter
  // consumers never see two dispatch paths for the same message. The
  // callback only exists to keep the placement alive on the native
  // manager (so it keeps polling).
  //
  // We hand Dart a monotonic Flutter-side subscription id and stash
  // the corresponding `Synapse.ShowToken` in a dictionary; the
  // companion `inAppUnregisterShow` looks the token up by id and
  // cancels it. This indirection exists because `Synapse.ShowToken`'s
  // `subscriptionId` is private — we can't surface it to Dart even
  // though the cross-SDK contract names a subscription handle.

  /// Monotonically-increasing Flutter-side subscription id. Allocated
  /// per `inAppShow` call; never reused. Atomic via the `tokensLock`.
  private var nextSubscriptionId: Int64 = 0
  private var tokensById: [Int64: Synapse.ShowToken] = [:]
  private let tokensLock = NSLock()

  func inAppShow(
    placement: String,
    completion: @escaping (Result<InAppShowTokenDto, Error>) -> Void
  ) {
    Task {
      let token = await Synapse.InApp.show(placement: placement) { _ in
        // No-op — the Dart umbrella dispatches per-placement via the
        // shared event stream's `inAppMessageReceived` envelope. The
        // callback is required by `Synapse.InApp.show` so the manager
        // tracks the placement; we don't need to do anything here.
      }

      let id = await MainActor.run { () -> Int64 in
        tokensLock.lock()
        defer { tokensLock.unlock() }
        nextSubscriptionId &+= 1
        let id = nextSubscriptionId
        tokensById[id] = token
        return id
      }

      completion(.success(InAppShowTokenDto(
        placement: placement,
        subscriptionId: id
      )))
    }
  }

  func inAppUnregisterShow(
    placement: String,
    subscriptionId: Int64,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    let token: Synapse.ShowToken? = {
      tokensLock.lock()
      defer { tokensLock.unlock() }
      return tokensById.removeValue(forKey: subscriptionId)
    }()
    // `cancel()` is idempotent; calling on the deinit-already-fired
    // path is a silent no-op. Safe to call here even if Dart racing
    // with two `dispose()` from different code paths drops the
    // entry from the dictionary first.
    token?.cancel()
    completion(.success(()))
  }

  func inAppGetActive(
    placement: String?,
    completion: @escaping (Result<[InAppMessageDto], Error>) -> Void
  ) {
    Task {
      let messages = await Synapse.InApp.getActive(placement: placement)
      completion(.success(messages.map(Self.encodeInAppMessage)))
    }
  }

  func inAppDismiss(
    messageId: String,
    reason: String?,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    Task {
      await Synapse.InApp.dismiss(messageId: messageId, reason: reason)
      completion(.success(()))
    }
  }

  func inAppMarkInteracted(
    messageId: String,
    ctaId: String,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    Task {
      await Synapse.InApp.markInteracted(messageId: messageId, ctaId: ctaId)
      completion(.success(()))
    }
  }

  func inAppRefresh(completion: @escaping (Result<Void, Error>) -> Void) {
    Task {
      await Synapse.InApp.refresh()
      completion(.success(()))
    }
  }

  // MARK: - Helpers

  /// Translate optional level-string into the SDK enum. Returns nil for
  /// unrecognised values so the caller can raise `invalid_argument`.
  private static func parseLogLevel(_ level: String?) -> LogLevel? {
    switch level?.lowercased() {
    case nil, "info":
      return .info
    case "debug":
      return .debug
    case "warning", "warn":
      return .warning
    case "error":
      return .error
    case "none", "off", "silent":
      return LogLevel.none
    default:
      return nil
    }
  }

  private static func logLevelString(_ level: LogLevel) -> String {
    switch level {
    case .debug: return "debug"
    case .info: return "info"
    case .warning: return "warning"
    case .error: return "error"
    case LogLevel.none: return "none"
    }
  }

  private static func permissionString(_ status: PushPermissionStatus) -> String {
    switch status {
    case .authorized: return "granted"
    case .denied: return "denied"
    case .provisional: return "provisional"
    case .ephemeral: return "ephemeral"
    case .notDetermined: return "notDetermined"
    }
  }

  /// Decode the JSON string Dart sends into the SDK's typed JSONValue map.
  /// Returns nil if the input is nil.
  private static func decodeTraits(_ json: String?) throws -> [String: JSONValue]? {
    guard let json = json, !json.isEmpty else { return nil }
    let data = Data(json.utf8)
    return try JSONDecoder().decode([String: JSONValue].self, from: data)
  }

  private static func encodeIdentity(_ result: IdentityResult) -> PyrxIdentityResult {
    PyrxIdentityResult(
      contactId: result.contactId.uuidString,
      path: result.path.rawValue,
      aliasedExternalId: result.aliasedExternalId,
      eventsReattributed: Int64(result.eventsReattributed),
      devicesReattributed: Int64(result.devicesReattributed)
    )
  }

  /// Map PyrxError cases to the wire-level error code set the Dart layer
  /// expects. Mirrors the RN bridge's translation table.
  private static func translate(_ error: Error) -> FlutterError {
    if let pyrxError = error as? PyrxError {
      switch pyrxError {
      case .notInitialized:
        return flutterError(code: "not_initialized", message: "Pyrx.shared.initialize has not completed")
      case .invalidConfig(let reason):
        return flutterError(code: "invalid_argument", message: reason)
      case .network(let info):
        return flutterError(code: "network_error", message: String(describing: info))
      case .keychainFailure(let status, let operation):
        return flutterError(code: "internal_error", message: "keychain failure on \(operation): OSStatus \(status)")
      case .alreadyInitialized:
        return flutterError(code: "invalid_argument", message: "Pyrx.shared.initialize was called twice with different configs")
      @unknown default:
        return flutterError(code: "internal_error", message: String(describing: pyrxError))
      }
    }
    return flutterError(code: "internal_error", message: error.localizedDescription)
  }

  private static func flutterError(code: String, message: String?) -> FlutterError {
    FlutterError(code: code, message: message, details: nil)
  }

  // MARK: - In-app messaging encoders

  /// Convert a published `InAppMessage` into the Pigeon-wire DTO.
  /// CTAs lose nothing; `customData` JSONValues are projected onto the
  /// `Map<String?, Object?>` Pigeon shape (same flattening as
  /// `pyrx_attrs` on push payloads). `expiresAt` is rendered as ISO-8601
  /// UTC — the same format the backend emits.
  static func encodeInAppMessage(_ message: InAppMessage) -> InAppMessageDto {
    return InAppMessageDto(
      id: message.id,
      messageId: message.messageId,
      placement: message.placement,
      title: message.title,
      body: message.body,
      imageUrl: message.imageUrl,
      ctas: message.ctas.map(encodeInAppCta),
      customData: message.customData.map(encodeCustomDataMap),
      expiresAt: message.expiresAt.map(iso8601.string(from:)),
      priority: Int64(message.priority)
    )
  }

  private static func encodeInAppCta(_ cta: InAppCta) -> InAppCtaDto {
    return InAppCtaDto(
      id: cta.id,
      label: cta.label,
      actionType: cta.actionType.rawValue,
      actionPayload: cta.actionPayload
    )
  }

  /// Project a typed JSONValue map onto the loosely-typed Pigeon
  /// `Map<String?, Object?>` shape. NSNull stands in for nil so the
  /// Pigeon codec round-trips the explicit-null case faithfully.
  private static func encodeCustomDataMap(_ map: [String: JSONValue]) -> [String?: Any?] {
    var out: [String?: Any?] = [:]
    for (k, v) in map {
      out[k] = encodeJSONValue(v)
    }
    return out
  }

  private static func encodeJSONValue(_ value: JSONValue) -> Any {
    switch value {
    case .null:
      return NSNull()
    case let .string(s):
      return s
    case let .int(i):
      return i
    case let .double(d):
      return d
    case let .bool(b):
      return b
    case let .array(arr):
      return arr.map { encodeJSONValue($0) }
    case let .object(obj):
      var dict: [String: Any] = [:]
      for (k, v) in obj { dict[k] = encodeJSONValue(v) }
      return dict
    }
  }

  private static let iso8601: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return f
  }()
}

File: packages/pyrx_synapse_android/android/src/main/kotlin/tech/pyrx/synapse/flutter/PyrxSynapseHostApiImpl.kt
/*
 * PyrxSynapseHostApiImpl.kt
 * pyrx_synapse_android — Pigeon HostApi implementation.
 *
 * Implements the Pigeon-generated PyrxSynapseHostApi interface by
 * forwarding every method to the published synapse-core / synapse-push
 * SDKs (Pyrx.* and PyrxPush.*). Every Pigeon callback runs through a
 * single per-plugin CoroutineScope so the suspend functions on Pyrx
 * (which serialize through internal Mutex) don't compete with each
 * other.
 *
 * Error mapping mirrors the iOS impl + the RN Android bridge:
 *
 *   PyrxError.NotInitialized       → "not_initialized"
 *   PyrxError.InvalidConfig        → "invalid_argument"
 *   PyrxError.Network              → "network_error"
 *   PyrxError.StorageFailure       → "internal_error"
 *   PyrxError.AlreadyInitialized   → "invalid_argument"
 *   <anything else>                → "internal_error"
 *
 * Permission flow
 * ---------------
 * Pyrx (Android) has no requestPushPermission method itself — Android
 * 13+ POST_NOTIFICATIONS is a runtime permission that requires an
 * Activity. The plugin owns this dance: pre-Android 13 returns
 * "granted" immediately; Android 13+ checks current permission state,
 * else calls ActivityCompat.requestPermissions and resolves the Pigeon
 * completion handler from onRequestPermissionsResult below.
 *
 * On configurations without a foreground Activity (the bridge attached
 * before the customer's MainActivity is up) the impl returns
 * "notDetermined" so the caller can retry from a screen.
 */

package tech.pyrx.synapse.flutter

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import java.util.UUID
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import tech.pyrx.synapse.LogLevel
import tech.pyrx.synapse.Pyrx
import tech.pyrx.synapse.PyrxConfig
import tech.pyrx.synapse.PyrxEnvironment
import tech.pyrx.synapse.PyrxError
import tech.pyrx.synapse.flutter.generated.FlutterError
import tech.pyrx.synapse.flutter.generated.PyrxDebugInfo as PigeonPyrxDebugInfo
import tech.pyrx.synapse.flutter.generated.PyrxIdentityResult as PigeonPyrxIdentityResult
import tech.pyrx.synapse.flutter.generated.PyrxInitArgs
import tech.pyrx.synapse.flutter.generated.PyrxPushPermissionResult
import tech.pyrx.synapse.flutter.generated.PyrxSynapseHostApi
import tech.pyrx.synapse.identity.IdentityResult
import tech.pyrx.synapse.network.JSONValue
import tech.pyrx.synapse.PyrxDebugInfo as NativePyrxDebugInfo

internal class PyrxSynapseHostApiImpl(
    private val appContext: Context,
) : PyrxSynapseHostApi {

    /// One scope per plugin instance. SupervisorJob means a failed
    /// dispatch doesn't cancel sibling work; Dispatchers.Default keeps
    /// us off the UI thread for SDK calls (the SDK is suspend-safe).
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    private val json = Json { ignoreUnknownKeys = true }

    /// Activity reference for permission requests, set by the plugin
    /// via setActivity() when ActivityAware fires.
    @Volatile
    private var activity: Activity? = null

    private var pendingPermissionCallback: ((Result<PyrxPushPermissionResult>) -> Unit)? = null

    internal fun setActivity(activity: Activity?) {
        this.activity = activity
    }

    // ----------------------------------------------------------------
    // Lifecycle
    // ----------------------------------------------------------------

    override fun initialize(
        args: PyrxInitArgs,
        callback: (Result<Unit>) -> Unit,
    ) {
        val workspaceUuid = runCatching { UUID.fromString(args.workspaceId) }
            .getOrElse {
                callback(failure("invalid_argument", "workspaceId must be a UUID v4 string"))
                return
            }

        val environment = when (args.environment.lowercase()) {
            "production", "live" -> PyrxEnvironment.PRODUCTION
            "sandbox", "staging", "test" -> PyrxEnvironment.SANDBOX
            else -> {
                callback(failure("invalid_argument", "environment must be one of: production, sandbox"))
                return
            }
        }

        val logLevel = parseLogLevel(args.logLevel) ?: LogLevel.INFO

        val config = PyrxConfig(
            workspaceId = workspaceUuid,
            apiKey = args.apiKey,
            environment = environment,
            baseUrl = args.baseUrl ?: PyrxConfig.DEFAULT_BASE_URL,
            logLevel = logLevel,
            sdkVariant = "flutter",
        )

        scope.launch {
            try {
                Pyrx.initialize(context = appContext, config = config)
                callback(Result.success(Unit))
            } catch (e: PyrxError) {
                callback(Result.failure(translate(e)))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    override fun setLogLevel(
        level: String,
        callback: (Result<Unit>) -> Unit,
    ) {
        val parsed = parseLogLevel(level)
        if (parsed == null) {
            callback(failure("invalid_argument", "level must be one of: debug, info, warning, error, none"))
            return
        }
        Pyrx.setLogLevel(parsed)
        callback(Result.success(Unit))
    }

    override fun debugInfo(callback: (Result<PigeonPyrxDebugInfo>) -> Unit) {
        scope.launch {
            try {
                val info = Pyrx.debugInfo()
                callback(Result.success(encodeDebugInfo(info)))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    // ----------------------------------------------------------------
    // Identity
    // ----------------------------------------------------------------

    override fun identify(
        externalId: String,
        traitsJson: String?,
        callback: (Result<PigeonPyrxIdentityResult>) -> Unit,
    ) {
        val traits = try {
            decodeTraits(traitsJson)
        } catch (e: Throwable) {
            callback(failure("invalid_argument", "traits is not valid JSON: ${e.message}"))
            return
        }
        scope.launch {
            try {
                val result = Pyrx.identify(externalId = externalId, traits = traits)
                callback(Result.success(encodeIdentity(result)))
            } catch (e: PyrxError) {
                callback(Result.failure(translate(e)))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    override fun alias(
        newExternalId: String,
        callback: (Result<PigeonPyrxIdentityResult>) -> Unit,
    ) {
        scope.launch {
            try {
                val result = Pyrx.alias(newExternalId = newExternalId)
                callback(Result.success(encodeIdentity(result)))
            } catch (e: PyrxError) {
                callback(Result.failure(translate(e)))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    override fun logout(callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                Pyrx.logout()
                callback(Result.success(Unit))
            } catch (e: PyrxError) {
                callback(Result.failure(translate(e)))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    // ----------------------------------------------------------------
    // Events
    // ----------------------------------------------------------------

    override fun track(
        eventName: String,
        propertiesJson: String?,
        callback: (Result<Unit>) -> Unit,
    ) {
        val properties = try {
            decodeTraits(propertiesJson)
        } catch (e: Throwable) {
            callback(failure("invalid_argument", "properties is not valid JSON: ${e.message}"))
            return
        }
        scope.launch {
            try {
                Pyrx.track(eventName = eventName, properties = properties)
                callback(Result.success(Unit))
            } catch (e: PyrxError) {
                callback(Result.failure(translate(e)))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    override fun screen(
        screenName: String,
        propertiesJson: String?,
        callback: (Result<Unit>) -> Unit,
    ) {
        val properties = try {
            decodeTraits(propertiesJson)
        } catch (e: Throwable) {
            callback(failure("invalid_argument", "properties is not valid JSON: ${e.message}"))
            return
        }
        scope.launch {
            try {
                Pyrx.screen(screenName = screenName, properties = properties)
                callback(Result.success(Unit))
            } catch (e: PyrxError) {
                callback(Result.failure(translate(e)))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    // ----------------------------------------------------------------
    // Push
    // ----------------------------------------------------------------

    override fun requestPushPermission(
        alert: Boolean,
        sound: Boolean,
        badge: Boolean,
        callback: (Result<PyrxPushPermissionResult>) -> Unit,
    ) {
        // Pre-Android 13: notification permission granted at install
        // when the customer's app manifest declares POST_NOTIFICATIONS
        // (typically pulled in by synapse-push's transitive
        // dependencies). Report granted unconditionally.
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            callback(Result.success(PyrxPushPermissionResult("granted")))
            return
        }
        val alreadyGranted = ContextCompat.checkSelfPermission(
            appContext,
            Manifest.permission.POST_NOTIFICATIONS,
        ) == PackageManager.PERMISSION_GRANTED
        if (alreadyGranted) {
            callback(Result.success(PyrxPushPermissionResult("granted")))
            return
        }
        val act = activity
        if (act == null) {
            // No foreground Activity to anchor the OS dialog. Caller
            // can re-request once their UI is mounted.
            callback(Result.success(PyrxPushPermissionResult("notDetermined")))
            return
        }
        pendingPermissionCallback = callback
        ActivityCompat.requestPermissions(
            act,
            arrayOf(Manifest.permission.POST_NOTIFICATIONS),
            PUSH_PERMISSION_REQUEST_CODE,
        )
    }

    override fun registerForPushNotifications(callback: (Result<Unit>) -> Unit) {
        // FCM auto-registers via PyrxMessagingService (installed by
        // PyrxPush.install at plugin attach). No-op on Android.
        callback(Result.success(Unit))
    }

    /**
     * Called from the plugin's ActivityAware permission listener.
     * Returns true if we consumed the result, false if it was for
     * another permission request.
     */
    internal fun onRequestPermissionsResult(
        requestCode: Int,
        grantResults: IntArray,
    ): Boolean {
        if (requestCode != PUSH_PERMISSION_REQUEST_CODE) return false
        val callback = pendingPermissionCallback ?: return true
        pendingPermissionCallback = null
        val granted = grantResults.isNotEmpty() &&
            grantResults[0] == PackageManager.PERMISSION_GRANTED
        val status = if (granted) "granted" else "denied"
        callback(Result.success(PyrxPushPermissionResult(status)))
        return true
    }

    // ----------------------------------------------------------------
    // Privacy
    // ----------------------------------------------------------------

    override fun setTrackingEnabled(
        enabled: Boolean,
        callback: (Result<Unit>) -> Unit,
    ) {
        scope.launch {
            try {
                Pyrx.setTrackingEnabled(enabled = enabled)
                callback(Result.success(Unit))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    override fun deleteUser(callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                Pyrx.deleteUser()
                callback(Result.success(Unit))
            } catch (e: PyrxError) {
                callback(Result.failure(translate(e)))
            } catch (e: Throwable) {
                callback(failure("internal_error", e.message ?: e.toString()))
            }
        }
    }

    // ----------------------------------------------------------------
    // Helpers
    // ----------------------------------------------------------------

    private fun parseLogLevel(level: String?): LogLevel? =
        when (level?.lowercase()) {
            null, "info" -> LogLevel.INFO
            "debug" -> LogLevel.DEBUG
            "warning", "warn" -> LogLevel.WARNING
            "error" -> LogLevel.ERROR
            "none", "off", "silent" -> LogLevel.NONE
            else -> null
        }

    private fun logLevelString(level: LogLevel): String = when (level) {
        LogLevel.DEBUG -> "debug"
        LogLevel.INFO -> "info"
        LogLevel.WARNING -> "warning"
        LogLevel.ERROR -> "error"
        LogLevel.NONE -> "none"
    }

    private fun decodeTraits(raw: String?): Map<String, JSONValue>? {
        if (raw.isNullOrEmpty()) return null
        // JSONValueSerializer in synapse-core handles per-value
        // discrimination; deserializing into the map directly relies
        // on the kotlinx.serialization Map<String, JSONValue> path.
        return json.decodeFromString(
            kotlinx.serialization.serializer<Map<String, JSONValue>>(),
            raw,
        )
    }

    private fun encodeIdentity(result: IdentityResult): PigeonPyrxIdentityResult =
        PigeonPyrxIdentityResult(
            contactId = result.contactId,
            path = result.path.name.lowercase(),
            aliasedExternalId = result.aliasedExternalId,
            eventsReattributed = result.eventsReattributed.toLong(),
            devicesReattributed = result.devicesReattributed.toLong(),
        )

    private fun encodeDebugInfo(info: NativePyrxDebugInfo): PigeonPyrxDebugInfo {
        // The Android PyrxDebugInfo exposes hasExternalId as a Boolean
        // (not the value itself) -- privacy by default. The Pigeon
        // contract carries externalId as String?, so we project to null
        // on Android and surface only the "has it" bit through a side
        // channel in PR-2's umbrella when the customer-facing debug
        // viewer needs it.
        return PigeonPyrxDebugInfo(
            sdkVersion = info.sdkVersion,
            platform = info.platform,
            initialized = info.initialized,
            workspaceId = info.workspaceId?.toString(),
            environment = info.environment,
            baseUrl = info.baseUrl,
            logLevel = logLevelString(info.logLevel),
            anonymousId = info.anonymousId,
            externalId = null,
            trackingEnabled = info.trackingEnabled,
            queueDepth = info.eventQueueDepth.toLong(),
            deviceTokenFingerprint = info.deviceTokenFingerprint,
        )
    }

    private fun translate(error: PyrxError): FlutterError = when (error) {
        is PyrxError.NotInitialized ->
            FlutterError("not_initialized", "Pyrx.initialize has not completed", null)
        is PyrxError.InvalidConfig ->
            FlutterError("invalid_argument", error.reason, null)
        is PyrxError.Network ->
            FlutterError("network_error", error.inner.toString(), null)
        is PyrxError.StorageFailure ->
            FlutterError("internal_error", "storage failure on ${error.operation}", null)
        is PyrxError.AlreadyInitialized ->
            FlutterError(
                "invalid_argument",
                "Pyrx.initialize was called twice with different configs",
                null,
            )
    }

    private fun <T> failure(code: String, message: String?): Result<T> =
        Result.failure(FlutterError(code, message, null))

    companion object {
        private const val PUSH_PERMISSION_REQUEST_CODE = 0x50_59_52_58 // "PYRX"
    }
}
